llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp

```diff
@@ -1649,12 +1649,19 @@ void DAGTypeLegalizer::ExpandFloatRes_FCEIL(SDNode *N,
 
 void DAGTypeLegalizer::ExpandFloatRes_FCOPYSIGN(SDNode *N,
                                                 SDValue &Lo, SDValue &Hi) {
-  ExpandFloatRes_Binary(N, GetFPLibCall(N->getValueType(0),
-                                        RTLIB::COPYSIGN_F32,
-                                        RTLIB::COPYSIGN_F64,
-                                        RTLIB::COPYSIGN_F80,
-                                        RTLIB::COPYSIGN_F128,
-                                        RTLIB::COPYSIGN_PPCF128), Lo, Hi);
+
+  assert(N->getValueType(0) == MVT::ppcf128 &&
+         "Logic only correct for ppcf128!");
+  SDLoc DL = SDLoc(N);
+  SDValue Tmp = SDValue();
+  GetExpandedFloat(N->getOperand(0), Lo, Tmp);
+
+  Hi = DAG.getNode(ISD::FCOPYSIGN, DL, Tmp.getValueType(), Tmp,
+                   N->getOperand(1));
+  // a double-double is Hi + Lo, so if Hi flips sign, so must Lo
+  Lo = DAG.getSelectCC(DL, Tmp, Hi, Lo,
+                       DAG.getNode(ISD::FNEG, DL, Lo.getValueType(), Lo),
+                       ISD::SETEQ);
```

Comment:
Using SETEQ for FP types is weird, since it's the case that may be oeq or ueq. Can this break the handling of nan sign bits?

Can this just be emitted as another copysign?

Reply:
Hmm. Because the rule is Hi + Lo == "the real float", if one is NaN then logically they both are, aren't they?

Though, to be clear:
- If the double-double is NaN or Infinity, the Hi value must encode NaN or Infinity.
- If the double-double is Infinity, then Lo must not be NaN.
- Otherwise, the value of Lo is not significant.

Reply:
Basically, the answer of "how does `ppc_fp128` work for this" is "mmmostly by not really caring too hard, honestly".

Which doesn't mean I am against using the SETUEQ version. If I understand how the `getSelectCC` API works it does _sound_ more correct?

Reply:
Unlike a true floating point operation, the "not significant" bits of a nan shouldn't change. It would be better if we weren't flipping the bit of the insignificant half here (or for any of fneg/fabs/fcopysign)